File: app/src/main/java/com/coderstory/purify/module/ThemePatcher.java
package com.coderstory.purify.module;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;

import com.coderstory.purify.plugins.IModule;
import com.coderstory.purify.utils.XposedHelper;

import de.robv.android.xposed.IXposedHookZygoteInit;
import de.robv.android.xposed.XC_MethodHook;
import de.robv.android.xposed.XC_MethodReplacement;
import de.robv.android.xposed.XposedBridge;
import de.robv.android.xposed.callbacks.XC_InitPackageResources;
import de.robv.android.xposed.callbacks.XC_LoadPackage;

import static com.coderstory.purify.config.Misc.isEnable;


public class ThemePatcher extends XposedHelper implements IModule {


    @Override
    public void handleInitPackageResources(XC_InitPackageResources.InitPackageResourcesParam resparam) {

    }

    @Override
    public void handleLoadPackage(XC_LoadPackage.LoadPackageParam lpparam) {

        if (!isEnable()) {
            return;
        }
        // hook 框架层的root检测
        if (("android".equals(lpparam.packageName))) {
            XposedBridge.log("阶段1");
            XposedBridge.log("hook次数" + hookAllMethods("com.android.server.DeviceStateService", lpparam.classLoader, "doCheckState", new XC_MethodHook() {
                @Override
                protected void afterHookedMethod(MethodHookParam param) throws Throwable {
                    super.afterHookedMethod(param);

                    int code = (int) param.args[0];
                    if (code == 1) {
                        param.setResult(0);
                    }
                }
            }));
        }

        // 主题和谐
        if (lpparam.packageName.equals("com.meizu.customizecenter") && prefs.getBoolean("enabletheme", false)) {
            if (lpparam.packageName.equals("com.meizu.customizecenter")) {
                // 拦截开机自启广播
                findAndHookMethod("com.meizu.customizecenter.admin.receiver.BootBroadcastReceiver", lpparam.classLoader, "onReceive", Context.class, Intent.class, XC_MethodReplacement.returnConstant(null));

                // 拦截试用服务
                findAndHookMethod("com.meizu.customizecenter.manager.managermoduls.font.FontTrialService", lpparam.classLoader, "onStartCommand", Intent.class, int.class, int.class, XC_MethodReplacement.returnConstant(0));
                findAndHookMethod("com.meizu.customizecenter.manager.managermoduls.theme.ThemeTrialService", lpparam.classLoader, "onStartCommand", Intent.class, int.class, int.class, XC_MethodReplacement.returnConstant(0));

                //device_states | doCheckState
                // 8.2.4
                //findAndHookMethod("com.meizu.creator.commons.utils.DeviceUtils", lpparam.classLoader, "isPhoneRooted", Context.class, XC_MethodReplacement.returnConstant(false));
                //findAndHookMethod("com.meizu.creator.commons.utils.reflect.ReflectUtils", lpparam.classLoader, "doCheckRootState", Context.class, XC_MethodReplacement.returnConstant(false));
                //findAndHookMethod("com.meizu.creator.commons.utils.reflect.ReflectUtils", lpparam.classLoader, "doCheckRootState", Context.class, XC_MethodReplacement.returnConstant(false));
                //8.3.6
                //findAndHookMethod("com.meizu.customizecenter.manager.utilstool.a.b", lpparam.classLoader, "e", Context.class, XC_MethodReplacement.returnConstant(false));
                findAndHookMethod("com.meizu.net.lockscreenlibrary.manager.utilstool.baseutils.Utility", lpparam.classLoader, "isRoot", Context.class, XC_MethodReplacement.returnConstant(false));
                findAndHookMethod("com.meizu.statsapp.v3.lib.plugin.f.b", lpparam.classLoader, "h", Context.class, XC_MethodReplacement.returnConstant(false));
                //findAndHookMethod("com.meizu.customizecenter.manager.utilstool.a.b", lpparam.classLoader, "e", Context.class, XC_MethodReplacement.returnConstant(0));
                // com.meizu.advertise.plugin
                hookAllMethods("com.meizu.advertise.api.AdManager", lpparam.classLoader, "install", XC_MethodReplacement.returnConstant(null));

                //resetToSystemTheme
                findAndHookMethod("com.meizu.customizecenter.manager.managermoduls.theme.common.b", lpparam.classLoader, "c", XC_MethodReplacement.returnConstant(true));
                findAndHookMethod("com.meizu.customizecenter.manager.managermoduls.theme.common.b", lpparam.classLoader, "b", XC_MethodReplacement.returnConstant(true));

                /**
                 *
                 public void a(boolean arg4, boolean arg5) {
                 com.meizu.customizecenter.manager.utilstool.c.b.b(this.b, "startFontRestoreService");
                 Intent v0 = new Intent(this.d, FontRestoreService.class);
                 v0.putExtra(com.meizu.customizecenter.model.a.a$g.h.a(), this.e().a());
                 v0.putExtra("is_go_to_pay_font", arg5);
                 v0.putExtra("is_restore_last_key", arg4);
                 com.meizu.customizecenter.manager.utilstool.systemutills.a.a.a(this.d, v0);
                 }
                 */
                // 7.5
                findAndHookMethod("com.meizu.customizecenter.manager.managermoduls.font.k", lpparam.classLoader, "a", Context.class, String.class, long.class, XC_MethodReplacement.returnConstant(null));

                /**
                 *  public void k() {
                 *         long v0 = SystemClock.elapsedRealtime() - this.n();
                 *         com.meizu.customizecenter.manager.utilstool.c.b.e(this.b, "checkTrialFontWhenAppStart_interval:" + v0 / 1000 + " second");
                 *         if(!this.f(v0)) {
                 *             this.g();
                 *         }
                 *     }
                 */
                findAndHookMethod("com.meizu.customizecenter.manager.managermoduls.font.g", lpparam.classLoader, "k", XC_MethodReplacement.returnConstant(null));
                findAndHookMethod("com.meizu.customizecenter.manager.managermoduls.font.k", lpparam.classLoader, "k", XC_MethodReplacement.returnConstant(null));

                //"checkTrialFont:!isUsingTrialFont() Context context, String str, long j
                findAndHookMethod("com.meizu.customizecenter.manager.managermoduls.font.k", lpparam.classLoader, "a", Context.class, String.class, long.class, XC_MethodReplacement.returnConstant(null));

                // 7.5
                Class<?> themeContentProvider = findClass("com.meizu.customizecenter.manager.utilsprefs.dbprefs.dao.ThemeContentProvider", lpparam.classLoader);
                if (themeContentProvider == null) {
                    //8.3.6
                    themeContentProvider = findClass("com.meizu.customizecenter.manager.utilshelper.dbhelper.dao.ThemeContentProvider", lpparam.classLoader);
                }
                //主题混搭 ThemeContentProvider query Unknown URI
                findAndHookMethod(themeContentProvider, "query", Uri.class, String[].class, String.class, String[].class, String.class, new XC_MethodHook() {
                    @Override
                    protected void beforeHookedMethod(XC_MethodHook.MethodHookParam param) throws Throwable {
                        Object[] objs = param.args;
                        String Tag = "(ITEMS LIKE";
                        String Tag2 = "%zklockscreen;%";
                        String Tag3 = "%com.meizu.flyme.weather;%";
                        boolean result = false;
                        for (Object obj : objs) {
                            if (obj instanceof String && (((String) obj).contains(Tag) || obj.equals(Tag2) || obj.equals(Tag3))) {
                                result = true;
                            }
                        }
                        if (result) {
                            for (Object obj : objs) {
                                if (obj instanceof String[]) {
                                    for (int j = 0; j < ((String[]) obj).length; j++) {
                                        if (((String[]) obj)[j].contains("/storage/emulated/0/Customize/Themes")) {
                                            ((String[]) obj)[j] = "/storage/emulated/0/Customize%";
                                        } else if (((String[]) obj)[j].contains("/storage/emulated/0/Customize/TrialThemes")) {
                                            ((String[]) obj)[j] = "NONE";
                                        }
                                    }
                                }
                            }
                        }
                        super.beforeHookedMethod(param);
                    }
                });
            }
        }
    }


    @Override
    public void initZygote(IXposedHookZygoteInit.StartupParam startupParam) {

    }


}

File: app/src/main/java/com/coderstory/flyme/patchModule/ThemePatcher.kt
package com.coderstory.flyme.patchModule


import android.content.*
import android.net.Uri
import com.coderstory.flyme.tools.XposedHelper
import com.coderstory.flyme.xposed.IModule
import de.robv.android.xposed.IXposedHookZygoteInit.StartupParam
import de.robv.android.xposed.XC_MethodHook
import de.robv.android.xposed.XC_MethodReplacement
import de.robv.android.xposed.XposedBridge
import de.robv.android.xposed.callbacks.XC_InitPackageResources.InitPackageResourcesParam
import de.robv.android.xposed.callbacks.XC_LoadPackage.LoadPackageParam


class ThemePatcher : XposedHelper(), IModule {
    override fun handleInitPackageResources(respray: InitPackageResourcesParam) {}
    override fun handleLoadPackage(param: LoadPackageParam) {

        // 主题和谐
        if (param.packageName == "com.meizu.customizecenter" && prefs.getBoolean("enabletheme", false)) {
            if (param.packageName == "com.meizu.customizecenter") {
                XposedBridge.log("开始hook主题")
                // 开始试用主题
                findAndHookMethod(
                    "com.meizu.customizecenter.manager.managermoduls.theme.f",
                    param.classLoader,
                    "N",
                    String::class.java,
                    String::class.java,
                    Int::class.javaPrimitiveType,
                    Int::class.javaPrimitiveType,
                    XC_MethodReplacement.returnConstant(null)
                )

                // 拦截开机自启广播
                findAndHookMethod(
                    "com.meizu.customizecenter.admin.receiver.BootBroadcastReceiver",
                    param.classLoader,
                    "onReceive",
                    Context::class.java,
                    Intent::class.java,
                    XC_MethodReplacement.returnConstant(null)
                )

                // 拦截试用服务
                findAndHookMethod(
                    "com.meizu.customizecenter.manager.managermoduls.font.FontTrialService",
                    param.classLoader,
                    "onStartCommand",
                    Intent::class.java,
                    Int::class.javaPrimitiveType,
                    Int::class.javaPrimitiveType,
                    XC_MethodReplacement.returnConstant(2)
                )
                findAndHookMethod(
                    "com.meizu.customizecenter.manager.managermoduls.theme.ThemeTrialService",
                    param.classLoader,
                    "onStartCommand",
                    Intent::class.java,
                    Int::class.javaPrimitiveType,
                    Int::class.javaPrimitiveType,
                    object : XC_MethodReplacement() {
                        @Throws(Throwable::class)
                        override fun replaceHookedMethod(param: MethodHookParam): Any {
                            return 2
                        }
                    })

                //device_states | doCheckState
                // 8.0.23
                findAndHookMethod(
                    "com.meizu.customizecenter.k.c.a.c",
                    param.classLoader,
                    "w",
                    Context::class.java,
                    XC_MethodReplacement.returnConstant(2)
                )
                findAndHookMethod(
                    "com.meizu.customizecenter.l.e.a.e",
                    param.classLoader,
                    "v1",
                    Context::class.java,
                    XC_MethodReplacement.returnConstant(2)
                )
                findAndHookMethod(
                    "com.meizu.net.lockscreenlibrary.manager.utilstool.baseutils.Utility",
                    param.classLoader,
                    "isRoot",
                    Context::class.java,
                    XC_MethodReplacement.returnConstant(false)
                )
                findAndHookMethod(
                    "com.meizu.statsapp.v3.lib.plugin.f.b",
                    param.classLoader,
                    "n",
                    Context::class.java,
                    XC_MethodReplacement.returnConstant(false)
                )

                //resetToSystemTheme
                // findAndHookMethod("com.meizu.customizecenter.manager.managermoduls.theme.common.b", lpparam.classLoader, "c", XC_MethodReplacement.returnConstant(true));
                // 8.0.23
                findAndHookMethod(
                    "com.meizu.customizecenter.manager.managermoduls.theme.j.b",
                    param.classLoader,
                    "e",
                    XC_MethodReplacement.returnConstant(true)
                )
                findAndHookMethod(
                    "com.meizu.customizecenter.manager.managermoduls.theme.j.b",
                    param.classLoader,
                    "I",
                    XC_MethodReplacement.returnConstant(true)
                )
                findAndHookMethod(
                    "com.meizu.customizecenter.manager.managermoduls.theme.j.b",
                    param.classLoader,
                    "I",
                    XC_MethodReplacement.returnConstant(true)
                )
                /**
                 *
                 * public void a(boolean arg4, boolean arg5) {
                 * com.meizu.customizecenter.manager.utilstool.c.b.b(this.b, "startFontRestoreService");
                 * Intent v0 = new Intent(this.d, FontRestoreService.class);
                 * v0.putExtra(com.meizu.customizecenter.model.a.a$g.h.a(), this.e().a());
                 * v0.putExtra("is_go_to_pay_font", arg5);
                 * v0.putExtra("is_restore_last_key", arg4);
                 * com.meizu.customizecenter.manager.utilstool.systemutills.a.a.a(this.d, v0);
                 * }
                 */
                // 7.5
                findAndHookMethod(
                    "com.meizu.customizecenter.manager.managermoduls.font.k",
                    param.classLoader,
                    "a",
                    Context::class.java,
                    String::class.java,
                    Long::class.javaPrimitiveType,
                    XC_MethodReplacement.returnConstant(null)
                )
                findAndHookMethod(
                    "com.meizu.customizecenter.manager.managermoduls.font.k",
                    param.classLoader,
                    "Y",
                    Boolean::class.javaPrimitiveType,
                    Boolean::class.javaPrimitiveType,
                    XC_MethodReplacement.returnConstant(null)
                )
                hookAllMethods(
                    "com.meizu.customizecenter.manager.managermoduls.font.k",
                    param.classLoader,
                    "Y",
                    XC_MethodReplacement.returnConstant(null)
                );

                /**
                 * public void k() {
                 * long v0 = SystemClock.elapsedRealtime() - this.n();
                 * com.meizu.customizecenter.manager.utilstool.c.b.e(this.b, "checkTrialFontWhenAppStart_interval:" + v0 / 1000 + " second");
                 * if(!this.f(v0)) {
                 * this.g();
                 * }
                 * }
                 */
                findAndHookMethod(
                    "com.meizu.customizecenter.manager.managermoduls.font.g",
                    param.classLoader,
                    "k",
                    XC_MethodReplacement.returnConstant(null)
                )
                findAndHookMethod(
                    "com.meizu.customizecenter.manager.managermoduls.font.k",
                    param.classLoader,
                    "k",
                    XC_MethodReplacement.returnConstant(null)
                )
                // 8.0.23
                findAndHookMethod(
                    "com.meizu.customizecenter.manager.managermoduls.font.k",
                    param.classLoader,
                    "b",
                    XC_MethodReplacement.returnConstant(null)
                )
                findAndHookMethod(
                    "com.meizu.customizecenter.manager.managermoduls.font.k",
                    param.classLoader,
                    "m",
                    XC_MethodReplacement.returnConstant(null)
                )

                findAndHookMethod(
                    "com.meizu.customizecenter.manager.managermoduls.font.k",
                    param.classLoader,
                    "m",
                    XC_MethodReplacement.returnConstant(null)
                )


                //"checkTrialFont:!isUsingTrialFont() Context context, String str, long j
                findAndHookMethod(
                    "com.meizu.customizecenter.manager.managermoduls.font.k",
                    param.classLoader,
                    "a",
                    Context::class.java,
                    String::class.java,
                    Long::class.javaPrimitiveType,
                    XC_MethodReplacement.returnConstant(null)
                )
                findAndHookMethod(
                    "com.meizu.customizecenter.manager.managermoduls.font.k",
                    param.classLoader,
                    "l",
                    Context::class.java,
                    String::class.java,
                    Long::class.javaPrimitiveType,
                    XC_MethodReplacement.returnConstant(null)
                )

                hookAllMethods(
                    "com.meizu.customizecenter.manager.managermoduls.font.k",
                    param.classLoader,
                    "l",
                    XC_MethodReplacement.returnConstant(null)
                )
                findAndHookMethod(
                    "com.meizu.customizecenter.manager.managermoduls.font.k",
                    param.classLoader,
                    "l",
                    Context::class.java,
                    String::class.java,
                    Long::class.javaPrimitiveType,
                    XC_MethodReplacement.returnConstant(null)
                )

                val themeContentProvider: Class<*> = findClass(
                    "com.meizu.customizecenter.manager.utilshelper.dbhelper.dao.ThemeContentProvider",
                    param.classLoader
                )
                //主题混搭 ThemeContentProvider query Unknown URI
                findAndHookMethod(
                    themeContentProvider,
                    "query",
                    Uri::class.java,
                    Array<String>::class.java,
                    String::class.java,
                    Array<String>::class.java,
                    String::class.java,
                    object : XC_MethodHook() {
                        @Throws(Throwable::class)
                        override fun beforeHookedMethod(param: MethodHookParam) {
                            val objs = param.args
                            val tag = "(ITEMS LIKE"
                            val tag2 = "%zklockscreen;%"
                            val tag3 = "%com.meizu.flyme.weather;%"
                        var result = false
                        for (obj in objs) {
                            if (obj is String && (obj.contains(tag) || obj == tag2 || obj == tag3)) {
                                result = true
                            }
                        }
                        if (result) {
                            for (obj in objs) {
                                if (obj is Array<*>) { // obj is String[]
                                    for (j in (obj as Array<String>).indices) {
                                        if (obj[j].contains("/storage/emulated/0/Customize/Themes")) {
                                            obj[j] = "/storage/emulated/0/Customize%"
                                        } else if (obj[j].contains("/storage/emulated/0/Customize/TrialThemes")) {
                                            obj[j] = "NONE"
                                        }
                                    }
                                }
                            }
                        }
                        super.beforeHookedMethod(param)
                    }
                    })

                // android 11 8.30.2
                findAndHookMethod(
                    "com.meizu.customizecenter.manager.utilshelper.restorehelper.ThemeRestoreService",
                    param.classLoader,
                    "onStartCommand",
                    Intent::class.java,
                    Int::class.javaPrimitiveType,
                    Int::class.javaPrimitiveType,
                    XC_MethodReplacement.returnConstant(0)
                )
                findAndHookMethod(
                    "com.meizu.customizecenter.manager.utilshelper.restorehelper.FontRestoreService",
                    param.classLoader,
                    "onStartCommand",
                    Intent::class.java,
                    Int::class.javaPrimitiveType,
                    Int::class.javaPrimitiveType,
                    XC_MethodReplacement.returnConstant(0)
                )
                findAndHookMethod(
                    "com.meizu.customizecenter.manager.utilshelper.scannerhelper.CustomizeScannerService",
                    param.classLoader,
                    "onStartCommand",
                    Intent::class.java,
                    Int::class.javaPrimitiveType,
                    Int::class.javaPrimitiveType,
                    XC_MethodReplacement.returnConstant(0)
                )
            }
        }
    }

    override fun initZygote(startupParam: StartupParam?) {
        TODO("Not yet implemented")
    }


}